---
title: "PrepCourse_R_Lab2"
author: 
- name: Author # write your name here!
  affiliation: Master program in Biotechnology
  email: corresponding@author.com
date: '`r format(Sys.time(), "%d %B, %Y")`' 
output: BiocStyle::html_document
---

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# install necessary packages in setup chunk
# for example:
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("BiocStyle", version = "3.8")

# load necessary packages here if you need
library(BiocStyle)
library(dplyr)

# In case you want to add python code chunk, you need to install `reticulate` package
#library(reticulate)
#use_python("/usr/local/bin/python")
```

# R Assignment2
This is a “R Markdown” file of assignment in R. You need to follow the instructions and write down some answers to the questions. Carefully read through the whole document, once you answer all the questions, you can `knit` the file in order to get the output html file. It is required that you send the html page to **Jing Xu** at <jing.xu@dkfz.de> no later than **March 15,2019** to get your credit points for the course.

## Objectives
In this assignment, you need to

    * search R functions on your own
    * check help page of R functions
    * do simple statisitics with R

## Contact
In case you have any questions, or further interests in doing internships and degree projects, you are always welcome to contact your tutors in the Applied Bioinformatics group (Prof. Benedikt Brors, B330, DKFZ).


<font size="4"> $\color{blue}{\text{Have fun with R now!}}$ </font> 
=================================================================

# Create R objects
1. Create a numeric vector of '113355' and repeat the series for 4 times, but __ONLY__ show the first 8 numbers of the series. Store your final result in a variable called __y__. You can do this in separate steps or in one step. Try both ways!     
__Hint:__ Look up the help page of the function.
```{r}
# write your answers here!
## in separate steps

## in one step
```

2. Extract even numbers from a numeric vector __x__ defined for you.  
__Hint:__ We didn't mention this in the lecture, you need to think what kind of feature do even numbers have (or odd numbers do not have)?
```{r}
x <- seq(1,50)
# write your answers here!

```

3. Simple lottery  
Write an R script to select a winning lotto number. We need to pick 5 lotto jackpots from a total of 50 jackpots, each ball has an unique number on it (numbers range from 1 to 50). Output the winning number in a variable called __winner__.  
__Hint__: you may find out that the numbers can be different every time from random sampling, but we can only have one winner! How to solve this problem? 
```{r}
# write your answers here!

```


# Descriptive statistics
Two dice game is the one of the oldest game in the gambling, you need to bet on the number of the sum of two 6-sided dices. In order to make the best guess, we need to find out the distribution of the sum,  

- show it in a histogram (density plot) 
- find out the most and the second frequently appeared numbers

It's a bit complicated if we calculate it manually, but R can help you with that! Write your own R code!  
__Hint__: You may need to get all possible combination of numbers first. If you use random sampling, how many times do you need to roll the dices to get a good approximation? 
```{r Dice game}
# write your answers here!


```


# Solve sequencing-related problem
Now you have learned bash, python and R, how to solve step2 from Day1 with R? 
* Extract rows of **the first 100 lines** with the comment lines and save it in a txt file in your working directory with bash. You may try to use the Terminal with bash in Rstudio (but not required), in case you want to try to use it with Windows system, you need to go to store to install Linux system for Windows.  
* Read in this file into R  
* As you already know, this file has 9 columns but has no column names, add the following column names: "chr_name", "annotation_source", "feature_type", "START", "END", "score", "strand", "genomic_phase_CDS", "INFO"  
* Solve problems in step2 and anwswer additional questions.  

__Hint__: 1. You may need to work with list, check `sapply` function 2. You may try `dplyr` package

- Q1. What kinds of "feature types" exist in this file?
```{r feature types}
# write your answers here!

```

- Q2. How many kinds of "+" and "-" strands are there in this file?
```{r strands}
# write your answers here!

```

- Q3. Solve problems in step2 from Day1 with R and output the head and tail of the final table

```{r solve problems in step2}
# write your answers here!

```

Give head of your final table
```{r head}
# write your answers here!

```

Give tail of your final table
```{r tail}
# write your answers here!

```


- Q4. Export your data to a desired location

* Export to your final table to current working directory 
```{r echo=TRUE,eval=FALSE}
# write your answers here!

```

* Create a subdirectory under your working directory and export the table there
```{r echo=TRUE,eval=FALSE}
# write your answers here!

```


In case you want to use python to solve the problems, you can write python code here! But you need to install `reticulate` package first, see descriptions in the setup chunk! 

If you decide to you python, remove all characters after "python" to form an executable python code chunk.
```{python, include = FALSE, eval=FALSE}
print("Hello")
```



# Bonus: exploring heatmaps
This is a bonus question, you don't need to answer it if you don't want or you can't. In case of these situations, just write whatever functional R code in the question code chunk in order to get final output. 

Heatmaps are very commonly used in genomics, in this exercise, we'll build a demo data to create some simple heatmaps. You need to install `RColorBrewer` and `pheatmap` packages first, load them in the setup chunk. 

Run the demo chunk in the console to see the demo heatmap
```{r heatmap demo}
# Generate some data, suppose you did 4 experiments and got the mean values of expression on a number of genes
test = matrix(rnorm(200), 20, 4)
test[1:10, seq(1, 4, 2)] = test[1:10, seq(1, 4, 2)] + 3
test[11:20, seq(2, 4, 2)] = test[11:20, seq(2, 4, 2)] + 2
colnames(test) = paste("Test", 1:4, sep = "")
rownames(test) = paste("Gene", 1:20, sep = "")
head(test)

# Generate column annotations, assuming you have two groups:
annotation = data.frame(Group1 = factor(1:4 %% 2 == 0, labels = c("Class1", "Class2"))
                        , Group2 = 1:4)
rownames(annotation) = paste("Test", 1:4, sep = "")

pheatmap(test, annotation = annotation) # auto color

```

If you want to specify colors in the annotation labels (aka. color group1 and 2 in different color sets), what code would you need?
```{r}
# Specify colors of annotation labels

```


# Document for reproducibility
It's always a good practice to output the versions of R and all packages you used for generating the file. Sometimes the script will not work when running with a different R or package version. Reproducibility is highly important in programming as well as in wet labs!

* Which R function is intended to print version information about R? Check `sessionInfo()` function.

This is the last section of the assignment, if you have answered all the questions, you can `knit` this Rmd file now. You won't be able to do so if there's any error or unanswered questions left in this file.

You may get some warnings from Rmarkdown compiling, but this is not because your code it wrong. As long as you get the html file, you are good to hand it in.
```{r}
sessionInfo() 
```


